Migrate all notebooks to API V1

gaborcselle opened the pull request:
## Summary
Migrates all remaining V0.* calls to Open AI API V1.

**Approach:** The approach can be seen in [these commits](https://github.com/openai/openai-cookbook/compare/main...gaborcselle:openai-cookbook:migrate-to-api-v1-with-helper). It consists of 4 steps:

1. Create a helper script, [ipynb_migration_helper.py](https://github.com/openai/openai-cookbook/blob/8dce2186360247c66da4dc089d6bc470210f2140/ipynb_migration_helper.py) 
2. Use to [ipynb_migration_helper.py](https://github.com/openai/openai-cookbook/blob/8dce2186360247c66da4dc089d6bc470210f2140/ipynb_migration_helper.py) **extract all code cells from all notebooks**. This is necessary because `openai migration` does not apply its migrations to `.ipynb` files.
3. Run `openai migration` which applies [Grit's OpenAI V1 migrations](https://app.grit.io/migrations/new/openai) to all the cells from the notebooks.
4. Remerge all migrated cells with [ipynb_migration_helper.py](https://github.com/openai/openai-cookbook/blob/8dce2186360247c66da4dc089d6bc470210f2140/ipynb_migration_helper.py) using the command line option `--reinsert`
5. Delete all temporary files (extracted cell files) and the helper script.

## Motivation

We have been migrating this repo to V1 piecemeal. This is because the `openai migrate` command and the [migration tools provided by Grit](https://app.grit.io/migrations/new/openai) that were referenced in the [V1.0.0 migration guide](https://github.com/openai/openai-python/discussions/742) do not support `.ipynb` files.

I have gone ahead and taken the approach outlined above, with the helper script to extract and reinsert cells, in order to take care of all remaining migrations in a scalable manner.

## Testing
I have not tested each .ipynb notebook affected, but am rather relying on the correctness of the `openai migrate` command.

# Squash and Merge
*Maintainers:* If you decide to merge this, please remember to [Squash and Merge](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges) into a single commit, so the helper script and temporary files created by it will not become part of the record. (Apologies if I'm stating the obvious.)

I hope this is helpful, and happy to answer any questions. Let me know if you have any feedback!

gaborcselle commented:
Quick update - 'openai update' missed a bunch of items, I just manually went through them to fix them.

kristapratico commented:
Everything under azure/archive is intended to show how to use the Azure API with openai<1.X. The Azure examples for >=1.X are already migrated under azure/

gaborcselle commented:
Thanks for the guidance @kristapratico, I will revert the azure/archive changes, and will ping you when that is complete

gaborcselle commented:
Thanks again @kristapratico, I have reverted the changes to azure/archive. Appreciate the pointer!

logankilpatrick commented:
This is great! It looks like it removes the formatting in many cases, is it possible to retain that? Also, it would be nice to follow the new pattern of instantiating a client.

ARajgor commented:
### API Deprecation and response format

1) Many notebooks still use Completion API. it's legacy and not going to work after January 4th, 2024. upgrade to client.chat.completions.create()
2) Response data is also changed. upgrade to response.choices[0].message.content

gaborcselle commented:
@logankilpatrick Thanks for the pointer, I'll go through and will make sure the formatting is maintained. ETA 1 day.

gaborcselle commented:
@ARajgor Thanks for the pointers! I'm not sure I understand the point you are making - this PR updates the notebooks to address (1) and (2).

ARajgor commented on examples/Backtranslation_of_SQL_queries.py:
```diff
@@ -24,17 +26,15 @@ def get_candidates(
-        stop=stop,
-        n=n,
-    )
+    response = client.completions.create(engine=engine,
```
Here. This code is not gonna work. completion.create doesn't have an engine parameter. 
completions API. soos will be deprecated. 
use client.chat.completions.create()

ARajgor replied:
the whole function should be upgraded to client.chat.completions.create(). Many notebooks still use this old function. so after a few months, I guess it's not gonna work.

gaborcselle replied:
I see, this may have been an error in the migration script. I'll double check occurences of this. Thank you for the pointer @ARajgor!

gaborcselle commented:
Quick update @ARajgor and @logankilpatrick - I made some progress today in addressing your feedback.

Not quite there, will need some more time with this PR, hoping for later today or tomorrow.

My goal is to manually go through and review each notebook that is changed, but I haven't had time to do so yet.

Thanks again!

gaborcselle commented:
@logankilpatrick and @ARajgor - please take another look

I went through and manually checked each notebook for formatting. I also added the new client instantiation style to a bunch (but not all) of the existing notebooks, and fixed some smaller issues.

Thanks for taking another look, and let me know what you think!

ARajgor commented on examples/Search_reranking_with_cross-encoders.ipynb:
```diff
@@ -245,7 +245,7 @@
         "@retry(wait=wait_random_exponential(min=1, max=40), stop=stop_after_attempt(3))\n",
         "def document_relevance(query, document):\n",
-        "    response = openai.Completion.create(\n",
+        "    response = openai.chat.completions.create(\n",
```
- import openai, upgrade to new version and change to cilent variable in the file.
- update the model to gpt-4, gpt-4 turbo, gpt-3.5-turbo

- there is no prompt parameter it's message

- response format.

- maybe check the whole file if any other change is needed.

ARajgor commented on examples/Unit_test_writing_using_a_multi-step_prompt_with_older_completions_API.ipynb:
It's for older models but I don't see any use-case in the near future because of API endpoint deprecation. that text-davinci-002 and code-davinci-002 is not gonna available.

gaborcselle replied:
Agreed, I went ahead and deleted it.

ARajgor commented on examples/azure/archive/chat.ipynb:
You update the azure/archieve files. it specifically install version <1.0.0 so chat.completions is not available. revert back to originals.

ARajgor commented on examples/azure/archive/chat_with_your_own_data.ipynb:
revert back to original.

ARajgor commented on examples/azure/archive/functions.ipynb:
revert back to original.

ARajgor commented on examples/book_translation/translate_latex_book.ipynb:
```diff
@@ -224,14 +224,14 @@
     "        max_tokens=1500,\n",
     "    )\n",
-    "    result = response['choices'][0]['text'].strip()\n",
+    "    result = response.choices[0].text.strip()\n",
```
response['choices'][0]['message']['content']

gaborcselle replied:
changed to response.choices[0].message.content.strip()

ARajgor commented on examples/book_translation/translate_latex_book.ipynb:
```diff
@@ -224,14 +224,14 @@
     "{sample_translation[1]}\n",
     "'''\n",
-    "    response = openai.Completion.create(\n",
+    "    response = openai.chat.completions.create(\n",
```
import openai
upgrade to new version and change to cilent variable in the file.

ARajgor commented on examples/book_translation/translate_latex_book.ipynb:
```diff
@@ -224,14 +224,14 @@
+    "    result = response.choices[0].text.strip()\n",
     "    result = result.replace('\"\"\"', '') # remove the double quotes, as we used them to surround the text\n",
     "    return result\n",
     "print(translate_chunk(chunks[800], engine='text-davinci-002', dest_language='English'))"
```
update the model in engine parameter and where this translate_chunk function use

ARajgor reviewed (commented):
I only review 27 files. remaining still need a look. major bug I found in many files is 
import openai. it should be updated to a newer client version. please check into other files as well. If I made some mistakes in the review, feel free to say.

gaborcselle commented:
Thanks for the in-depth review @ARajgor, will make sure to address the comments, aiming for later this weekend!

gaborcselle commented on examples/Embedding_Wikipedia_articles_for_search.ipynb:
Added client = openai.OpenAI() and changed openai call to use client.

gaborcselle commented on examples/Entity_extraction_for_long_documents.ipynb:
```diff
@@ -139,7 +139,7 @@
     "    presence_penalty=0\n",
     "    )\n",
-    "    return \"1.\" + response['choices'][0]['text']"
+    "    return \"1.\" + response.choices[0].text"
```
Surely you mean response.choices[0].message.content :-)

gaborcselle commented on examples/How_to_call_functions_for_knowledge_retrieval.ipynb:
This notebook actually uses the requests to call https://api.openai.com/v1/chat/completions directly.

IMHO we should consider removing the notebook entirely.

gaborcselle commented on examples/Multiclass_classification_for_transactions.ipynb:
I've made these changes, but this notebook seems super out-of-date in general ...

gaborcselle commented on examples/Named_Entity_Recognition_to_enrich_text.ipynb:
```diff
@@ -440,7 +439,7 @@
+    "    response = openai.chat.completions.create(\n",
     "        model=\"gpt-3.5-turbo-0613\",\n",
     "        messages=messages,\n",
     "        functions=generate_functions(labels),\n",
```
I think this update falls slightly outside the scope, but happy to include this if you prefer. For now I've left a TODO with the link you included. (Thanks for the pointer!)

gaborcselle commented:
Thank you @ARajgor for your detailed and comprehensive feedback. I've made the requested changes.

I deleted `examples/Unit_test_writing_using_a_multi-step_prompt_with_older_completions_API.ipynb` per your suggestion.

@ARajgor and @logankilpatrick - please take another look! Thanks again for your suggestions and feedback.

gaborcselle commented:
@ARajgor @logankilpatrick Quick update that I've been going through the notebooks one-by-one and making minor corrections where needed.

ARajgor commented on examples/Backtranslation_of_SQL_queries.py:
It's outdated and very old logic. the functions are in respect to old models so the prompt needs to change. it directly passes the prompt into the messages parameter instead of the required list[dict] format. I think we should either remove the file or someone needs to change the logic.

gaborcselle replied:
I agree, will remove this

ARajgor commented on examples/Clustering_for_transaction_classification.ipynb:
```diff
@@ -399,7 +398,7 @@
     "        .values\n",
     "    )\n",
-    "    completion = client.chat.completions.create(\n",
+    "    response = openai.chat.completions.create(\n",
```
revert to client cause you already define the client in 2nd cell

gaborcselle replied:
Not sure what happened there, reverting this change.

ARajgor commented:
> @ARajgor @logankilpatrick Quick update that I've been going through the notebooks one-by-one and making minor corrections where needed. Lots of outdated code! Will continue tomorrow.

yes. I think we should check each file manually. That's how we make sure each file is working properly and has the latest changes.

gaborcselle commented:
Just merged in the latest changes from main. Will continue to go through the notebooks one by one.

ARajgor commented:
@gaborcselle I'll help you with that. let's have it plan out—a great task to work on in 2024 ;)

gaborcselle commented:
Thank you! Let me know how you want to split it up. Happy new year!

gaborcselle commented:
@ARajgor how do you want to split it up? I could make a Google Sheet for us that we put all the notebooks into, and then we check off the ones we've verified to work. Would that work for you?

gaborcselle commented:
@ARajgor I made this tracker spreadsheet - would this work to track the work? https://docs.google.com/spreadsheets/d/1DEARN1VobLr3pqHvEwubrmkzclSgefLGfuM76Lw1va0/edit#gid=1587694416

Want to DM me on X / Twitter with your email address so I can give you edit rights?

gaborcselle commented:
I made some progress and added some minor fixes, all recorded in here: https://docs.google.com/spreadsheets/d/1DEARN1VobLr3pqHvEwubrmkzclSgefLGfuM76Lw1va0/edit#gid=1587694416
Hoping to make some additional progress later this week.
(cc @ARajgor)

ARajgor commented:
## Update

I manually go through each file (except /third-party, /vector-databases) all are upgraded if necessary or modified based on the latest models. I run a few files without running out of my credits ;) 

Add me so that I can push my changes.

gaborcselle commented:
@ARajgor amazing, I invited you to my forked repo. Can you accept and push?

Thanks for updating the sheet as well, looking good! https://docs.google.com/spreadsheets/d/1DEARN1VobLr3pqHvEwubrmkzclSgefLGfuM76Lw1va0/edit#gid=1587694416

gaborcselle commented:
Thanks so much @ARajgor for pushing your code. Do you have an idea on how to solve the issues with `Code_search_using_embeddings.ipynb` and `./Embedding_long_inputs.ipynb` in the Google Sheet?

@logankilpatrick we'd love your guidance on the few notebooks marked "to deprecate" in the Google Sheet. Should we:
1. Remove them as part of this PR?
2. Add a note in the notebooks, at the top, that they are deprecated?
3. Keep as-is and submit a separate PR to deprecate?
Thank you!

ARajgor commented:
I already fixed these issues in the latest commit. test it again and tell me if it still shows the error.

1) As you mention in the sheet it has an issue with `df['code_embedding'] = df['code'].apply(lambda x: get_embedding(x, model='text-embedding-ada-002'))` That is because utils/embeddings_utils.py has an older function that's why.
2) for `openai.InvalidReqestError` it is replace with `openai.BadRequestError`.

gaborcselle commented:
Thanks @ARajgor, retesting now.

gaborcselle commented on examples/utils/embeddings_utils.py:
```diff
@@ -9,58 +9,57 @@
 from tenacity import retry, stop_after_attempt, wait_random_exponential
 
-import openai
+from openai import openAI
```
@ARajgor For me this yields and import error:
> ImportError: cannot import name 'openAI' from 'openai'

Should this OpenAI with a capital O

gaborcselle commented:
Quick update @ARajgor that I ported `Fine-tuned_classification.ipynb` from ada to babbage-002 and from the CLI interface to the V1 API. Training a babbage model and then using the legacy completions API feels like a remnant from a bygone era, so perhaps it's still time to deprecate that example? For now I left it in.

For the other two notebooks (`Multiclass_classification_for_transactions.ipynb` and `Recommendation_using_embeddings.ipynb`) I went ahead and marked them [Deprecated] at the very top. 

@logankilpatrick I think this PR is ready to go from our perspective, but we'd love your feedback / comments / guidance / critique, and we're happy to make changes. Let us know and thank you!

ARajgor commented:
@gaborcselle I think there is a mistake. `Recommendation_using_embedding.ipynb` doesn't use any Openai APIs it is a simple cosine-based recommendation example.

ARajgor commented on examples/Fine-tuned_classification.ipynb:
```diff
@@ -581,159 +516,158 @@
+   "source": [
+    "ft_model = fine_tune_results.fine_tuned_model\n",
+    "# note that thsi calls the legacy completions api\n",
+    "res = client.completions.create(model=ft_model, prompt=test['prompt'][0] + '\\n\\n###\\n\\n', max_tokens=1, temperature=0)\n",
```
It should be chat. completion. I guess many code blocks still rely on old API. it needs to be updated and the file should be tested. or else we should revert to the original and add a deprecation of the API notice on the top.

gaborcselle replied:
I did test this and it works - it relies on the legacy completions api which has no ".chat." and the prompt= param. I'll push a fix for my typo ("thsi").

I don't think it makes too much sense to have this file for much longer, but it does show you how you can use babbage-002 (instead of the deprecated ada) for classification.

ARajgor replied:
okay then.

ARajgor commented:
Can I suggest something, I reviewed some of the files from /third-party and /vector-dbs, it partially updated means if the import is correct then there is no client variable, something is wrong with the response format. so what I propose is let just revert those files to the current main version so that it still functions correctly. meanwhile, if we think we can fix those files then first we fix them one by one ensuring testing so that we avoid broken files merging into the main.

what's your point on this, @gaborcselle and @logankilpatrick

gaborcselle commented:
That sounds good. Do you want to go ahead and revert? If not, I can do it tomorrow.

gaborcselle commented:
Great catch @ARajgor, I fixed `Unit_test_writing_using_a_multi-step_prompt.ipynb` and updated the sheet: https://docs.google.com/spreadsheets/d/1DEARN1VobLr3pqHvEwubrmkzclSgefLGfuM76Lw1va0/edit#gid=1587694416

ARajgor commented:
Just to let you know, I just push my api key into commit. Thanks to GitGuardian I immediately take an action and revoked from the openai. ;)
Just check the spreadsheet and the files. Let's continue tomorrow.

gaborcselle commented:
I just set my OPENAI_API_KEY in a .env file at the root of the workspace and Visual Studio Code picks it up - much less risky that way.

What do you think is missing at this point? Just re-testing everything we've changed?

I've also messaged @logankilpatrick for feedback, would love to have him chime in.

ARajgor commented:
I guess everything is fine. There are two files in waiting. `./Multiclass_classification_for_transactions.ipynb` and `./Search_reranking_with_cross-encoders.ipynb` how to address these?

gaborcselle commented:
@ARajgor while we're waiting for feedback from @logankilpatrick, I'm going through and retesting all the notebooks. Do you want to go and do more in third_party?

ARajgor commented:
@gaborcselle I'll go through third-party and vector databases. I believe I can only do the AWS one, but let me double-check.

gaborcselle commented:
@ARajgor Thank you!

@logankilpatrick we'd love your feedback on this, let us know when you've had a chance to take a look.

gaborcselle commented:
@logankilpatrick, just wanted to check back in on this - sorry to keep pinging you. Our PR is ready, and fully tested: https://github.com/openai/openai-cookbook/pull/914

It migrates all of the various notebooks to V1. I think this would be helpful for a bunch of people trying to build stuff with the OpenAI API based off the cookbook examples. With this PR, they won't get stuck having to understand the old version the API vs the V1.

The PR is a clean merge, freshly updated and ready to go. We marked 2 .ipynb notebooks as deprecated, and could also remove them if you prefer.

Any guidance or feedback is most welcome, we are happy to make changes! 

Thank you, Gabor and @ARajgor

logankilpatrick commented:
This is huge! Thank you both for the hard work on getting this PR over the line.

gaborcselle commented:
> This is huge! Thank you both for the hard work on getting this PR over the line.

Thank you for reviewing and approving! Glad we got this to a conclusion!

vd4663rv commented:
I am walking through the steps, and get an error after being able to read the frames. I am unable to display the frame. 

618 frames read.
None
Traceback (most recent call last):
  File "c:\main.py", line 26, in <module>
    display_handle.update(Image(data=base64.b64decode(img.encode("utf-8"))))
    ^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'update'

Can someone help me on how to resolve this? thanks.
